File: java/retrieve heap/src/testsuites/Grass_Dl_Insert.java
package testsuites;

import java.util.Random;

public class Grass_Dl_Insert {
    static GNode create_list(int size){
        GNode root = null;
        GNode tl = root;
        for(int i = 0; i < size; i++){
            GNode new_node = new GNode();
            new_node.next = null;
            new_node.prev = null;

            if(root == null){
                root = new_node;
                tl = new_node;
            }else{
                tl.next = new_node;
                new_node.prev = tl;
                tl = new_node;
            }
        }

        return root;
    }

    static GNode dl_insert(GNode lst, GNode elt)
    {
        Random rand = new Random();
        //pre
        if (lst == null) {
            //post
            return elt;
        } else {
            int nondet = rand.nextInt(10) % 10 - 5;
            GNode curr = lst;
            while(nondet > 0 && curr.next != null)
            {
                //loop
                nondet = rand.nextInt(10) % 10 - 5;
                curr = curr.next;
            }
            GNode curr_next = curr.next;
            elt.next = curr_next;
            if (curr_next != null) {
                curr_next.prev = null;
            }
            curr.next = elt;
            elt.prev = curr;
            //post
            return lst;
        }
    }
    
    public static void main(String args[]){
        GNode lst = create_list(10);
        GNode elt = new GNode();
        elt.next = null;
        elt.prev = null;
        GNode res = dl_insert(lst, elt);
        dl_insert(null, elt);
    }
}

class GNode{
    GNode next;
    GNode prev;
}

File: benchmarks/python/grass_dl_insert.py
import random
import pprint

class node:
	def __init__(self, nxt = None, prev = None):
		self.next = nxt
		self.prev = prev

##################
# accessory func #
##################
def create_list(size):
	root = None
	tl = root
	for i in range(1, size):
		new_node = node()
		if root is None:
			root = new_node
			tl = root
		else:
			tl.next = new_node
			new_node.prev = tl
			tl = new_node
	return root

##################
#  end acc func  #
##################

def dl_insert(lst, elt):
	if lst is None:
		return elt
	else:
		nondet = 1
		curr = lst
		while nondet > 0 and curr.next is not None:
			nondet = random.randint(-2, 5)
			curr = curr.next
		curr_next = curr.next
		elt.next = curr_next
		if curr_next is not None:
			curr_next.prev = None
		curr.next = elt
		elt.prev = curr
		return lst

def main():
	root = create_list(10)
	elt = node()
	insert(root, elt)
	insert(None, elt)

if __name__=="__main__":
	main()
